### ISSUE TITLE
needles_range_loop unnecessarily adds .take(N) in many cases

### ISSUE THREAD

**Issue by gnzlbg** [rust-lang-nursery/rust-clippy#3033]
[Playground](https://play.rust-lang.org/?gist=d61644afd7b721a79da7bc8f84a55971&version=stable&mode=debug&edition=2015):

```rust
pub fn sum() -> u8 {
    let a: [u8; 3] = [1, 2, 3];
    let mut sum = 0;
    for i in 0..3 {
        sum += a[i];
    }
    sum
}
```

produces

```
warning: the loop variable `i` is only used to index `a`.
 --> src/main.rs:4:14
  |
4 |     for i in 0..3 {
  |              ^^^^
  |
  = note: #[warn(needless_range_loop)] on by default
  = help: for further information visit https://rust-lang-nursery.github.io/rust-clippy/v0.0.212/index.html#needless_range_loop
help: consider using an iterator
  |
4 |     for <item> in a.iter().take(3) {
  |         ^^^^^^    ^^^^^^^^^^^^^^^^
```

but if we apply the change:

```rust
pub fn sum() -> u8 {
    let a: [u8; 3] = [1, 2, 3];
    let mut sum = 0;
    for i in a.iter().take(3) {
        sum += i;
    }
    sum
}
```

we get a unnecessary `take(3)` that can actually be omitted since `a` only has `3` elements (and that's known at compile-time):

```rust
pub fn sum() -> u8 {
    let a: [u8; 3] = [1, 2, 3];
    let mut sum = 0;
    for i in a.iter() {
        sum += i;
    }
    sum
}
```
